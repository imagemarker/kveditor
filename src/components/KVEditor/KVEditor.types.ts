import { ReactElement } from 'react';

export interface KVEditConfigType {
  validateKey?: RegExp;
  theme?: string;
  idField?: string;
  stretchLabels?: boolean;
}
export interface KVEditorPropsType {
  items?: KVItemType[];
  rawObject?: any;
  options?: KVEditConfigType;
  onChange?: (items: any) => void;
}
export interface KVEditorType {
  (props: KVEditorPropsType): ReactElement;
}
export interface KVItemType {
  key: string;
  value: any;
  options?: KVItemOptionsType;
}
export interface KVItemOptionsType {
  immutable?: boolean; // non editable
  fixed?: boolean; // non removable
  nested?: boolean; // may include key/value list
  defaultValue?: any; // default value
}
export interface ValueEffectType {
  calculated?: any;
  validator?: any;
  parser?: any;
}
