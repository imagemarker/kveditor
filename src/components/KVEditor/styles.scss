.kv-editor {
  font-family: 'Source Sans Pro', sans-serif;
  font-weight: 400;
  margin: 3px;
  padding: 3px;
  border-radius: 3px;
  display: flex;
  flex-flow: column nowrap;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  min-width: 190px;

  &.light {
    color: #333;
    background-color: #FFF;
  }
  &.dark {
    color: #EEE;
    background-color: #444;
  }
}
