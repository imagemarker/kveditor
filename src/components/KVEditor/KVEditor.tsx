import React, { useEffect, useReducer } from 'react';
import { KVEditorType, KVItemType } from '@base/types';
import KVReducer from '../../hooks/KVReducer';
import KVItemEdit from '../KVItemEdit/KVItemEdit';
import KVItemView from '../KVItemView/KVItemView';
import './styles.scss';

const validateKey = new RegExp(/^[a-zA-Z][a-zA-Z0-9]*$/);

const KVEditor: KVEditorType = ({
  options = { theme: 'light', validateKey, typeNotation: 'string', nested: false }
}) => {
  const [state, dispatch] = useReducer(KVReducer, { items: [], keys: [] });

  useEffect(() => {
    const items = state.items.reduce(
      (acc: Record<string, unknown>, { key, value }) => ({ ...acc, [key]: value }),
      {}
    );
    console.log(items);
  }, [state]);

  return (
    <div className={`kv-editor ${options.theme}`}>
      {state.items.map((item: KVItemType) => (
        <KVItemView
          key={item.key}
          item={item}
          dispatch={dispatch}
          editorOptions={options}
          viewOptions={null}
        />
      ))}
      <KVItemEdit keys={state.keys} dispatch={dispatch} editorOptions={options} />
    </div>
  );
};

export default KVEditor;
