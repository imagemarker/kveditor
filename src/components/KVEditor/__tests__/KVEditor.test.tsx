import React from 'react';
import { render } from '@testing-library/react';
import { KVEditConfigType } from '../KVEditor.types';
import KVEditor from '../KVEditor';

describe('KVEditor', () => {
  const validateKey = new RegExp(/^[a-zA-Z][a-zA-Z0-9]*$/);

  const tree = (options: KVEditConfigType = { theme: 'light', validateKey }) =>
    render(<KVEditor options={options} />);

  test('renders properly', () => {
    const { asFragment } = tree();
    expect(asFragment()).toMatchSnapshot();
  });
});
