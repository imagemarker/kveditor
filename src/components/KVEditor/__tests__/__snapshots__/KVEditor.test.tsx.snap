// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KVEditor renders properly 1`] = `
<DocumentFragment>
  <div
    class="kv-editor light"
  >
    <form
      class="kv-item-edit light"
    >
      <div
        class="row"
      >
        <div
          class="input-key"
        >
          <input
            name="key"
            placeholder="key"
            type="text"
            value=""
          />
        </div>
        <div
          class="input-value"
        >
          <input
            name="value"
            placeholder="value"
            type="text"
            value=""
          />
        </div>
        <input
          class="input-submit"
          type="submit"
          value="Add"
        />
      </div>
    </form>
  </div>
</DocumentFragment>
`;
